### ISSUE TITLE
[Enhancement]: Reevaluate removal of title from checkout page

### ISSUE THREAD

**Issue by nielslange** [woothemes/woocommerce#48619]
> [!IMPORTANT]
> Before merging teh PR for this issue, consult with @PanosSynetos to verify whether adjustments are required for the SWW end-to-end tests.

### Describe the solution you'd like

In Issue #47529, we decided to remove the title from the checkout page because it seemed visually redundant. However, this change could impact the experience for screen reader users, who might not realise which page they are on.

To solve this, we could reintroduce the page title in the DOM, but keep it invisible to sighted users by using the `.screen-reader-text` CSS class. This approach will ensure the title is accessible to screen readers without affecting the visual layout of the page. For more information on this CSS class, can be found in the [WordPress Accessibility Handbook](https://make.wordpress.org/accessibility/handbook/markup/the-css-class-screen-reader-text/).

### Describe alternatives you've considered

Apart from reintroducing the page title in the DOM, we should also investigate alternatives, such as the use of aria elements, or the semantic use of HTML5 elements.

### Additional context

n/a

**Comment by ralucaStan** [comment]
One concern worth also addressing is that merchants can add a Header block on that page using an H1, which would result in having 2 x H1 tags on the page.

**Comment by nielslange** [comment]
Thanks for pointing this out, @ralucaStan. I haven't looked into this issue yet, so I cannot say what the best solution would be. As mentioned in the description above, we could add the page title as an aria element. That way, the screen reader reads the page title. And when a merchant would add an H1 headline, only one H1 headline would appear. However, the screen reader would read the title twice, once from the aria element and once from the H1 headline. That said, I haven't researched potential options yet.

**Comment by Manussakis** [comment]
I have started looking into this issue.

**Comment by Manussakis** [comment]
The solution I propose is to add the H1 title back to the Checkout template, but add the `screen-reader-text` class to it.
If merchants add a new H1 title, the `screen-reader-text` title should be hidden.
This can be done in two ways:

### 1 - CSS
If merchants add the H1 title from the Checkout page, it will be placed above the Checkout block. We can then hide the `screen-reader-text` title using the following CSS selector

```css
h1.screen-reader-text:has( + .wp-block-post-content h1) {
  display: none;
}
```

The problem is if merchants use the Appearance -> Editor -> Templates -> WooCommerce -> Page: Checkout to add the H1 title. In this case, the headings can be arranged in several orders and there is no way to create a CSS selector that will work for all cases.

### 2 - JS
Once the Checkout page loads, a script will get all H1 elements on the page, if there is more than one, the heading with the `screen-reader-text` class will be hidden.

@nielslange @ralucaStan Any thoughts on these approaches?

**Comment by vikrampm1** [comment]
@nielslange @ralucaStan can you review the approaches suggested above? This is currently blocked to proceed further.

**Comment by ralucaStan** [comment]
@nielslange could you take over this issue and help out @vikrampm1 to unblock this issue?

**Comment by nielslange** [comment]
Thanks for your proposed solutions, @Manussakis. I wonder how both options would look to the merchant in the editor. Will the merchant see the default Checkout title in both cases, even thought it will be hidden on the frontend? If so, would the merchant see two Checkout titles in the editor, so that one Checkout title appears in the frontend? 

Thinking out loud, I wonder if we should approach this issue from a different angle and utilise the settings. E.g. the title block has a setting called `Make title a link`.

<img width="1512" alt="Screenshot 2024-07-26 at 11 36 35" src="https://github.com/user-attachments/assets/fc147bfd-f759-4bba-9a5e-635b799838df">

Would it be an option to fork this block and add the setting `Show title on frontend`? By default, `Show title on frontend` would be disabled, similar to `Make title a link`. However, if wanted, a merchant can make the title visible in the frontend. This way, the title always appears in the editor. In the frontend, it's always available in the DOM, and if `Show title on frontend` is disabled, we simply add the [`screen-reader-text`](https://make.wordpress.org/accessibility/handbook/markup/the-css-class-screen-reader-text/) class.

<img width="1512" alt="Screenshot 2024-07-26 at 11 39 16" src="https://github.com/user-attachments/assets/1f459e1b-37ad-4b31-8a65-9379967dc19d">

---

If this is not an option, I'd be leaning towards the CSS approach, as I'm concerned about a performance impact of the JS solution. While the performance impact might be low, I believe we should try to avoid any performance impacts at this crucial stage of the checkout process.

**Comment by Manussakis** [comment]
@nielslange The title on the Checkout page uses the `core/post-title` block. Any update on its functionality will be available for all other titles that extend this block, like the one on the [Single Product template](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/templates/templates/blockified/single-product.html#L18).

We could create a new title block specifically for the Checkout page, but I believe this brings more complexity to the code.

I have looked again into the #47509 issue, which proposes the removal of the heading from the Checkout page, and I would like to share my opinion.

The justification for removing it was mainly UX improvements since the heading appears out of place. While I agree that the visual layout is not the best when the heading is visible, I believe that WordPress already gives merchants alternatives to work around it, like using the Advanced -> Additional CSS Class(es) field on the Inspector Control panel to add the `screen-reader-text` class (or any other class to align it).
![image](https://github.com/user-attachments/assets/c4491720-a395-4c4d-8690-2368d6bc61ea)

Besides that, the lack of alignment may vary per theme, so this is not an issue for all the cases.

My opinion is that not having the Checkout title before the form is worse than having it not visually aligned. As you mentioned, this is a crucial stage of the checkout process, and users need to be informed exactly what the page is about, regardless of whether they use assistive technologies or not.

I believe that WooCommerce should provide merchants with the best recommendation in terms of Accessibility, SEO, and UX, which is having a visible H1 title per page. Any customization should be merchants' responsibility.

Another question I have is:
Why are we only giving the merchants the ability to remove the title from the Checkout page?
I see that the heading on the Cart page is also misaligned. 
![image](https://github.com/user-attachments/assets/906a13f8-be2b-46d0-aaee-91256eff4b2b)

All that said, I vote to revert the changes that removed the title from the Checkout page and update its CSS on both the Cart and Checkout pages to better align them. Any other customization should be the merchants' responsibility.
![image](https://github.com/user-attachments/assets/03d9c9e4-add9-42e4-b5c1-3f518bf8626e)

I appreciate your patience in reading this long text. 😬

I am looking forward to hearing your thoughts.

**Comment by vikrampm1** [comment]
@nielslange can you able to review the comment above and provide feedback so we can proceed further? Thanks.

**Comment by wavvves** [comment]
Hi, after catching up on this , I see two main reasons for the change:

- We wanted to improve UX on mobile to shorten the page
- We wanted to prevent display problems across different themes

Both of these reasons seem valid, and indeed introduce improvements on the checkout experience. Now we have a a11y problem with screen readers, I think we can disregard the SEO as a problem since these pages shouldn't be indexed.

Having the merchant fix the problem by adding classes is not the ideal solution as it is not straightforward to many and could also be perceived as a nonexistent solution.
On the other hand, I don't believe adding a visibility toggle to `core/post-title` is going to be easily attainable with only this use case.

The solution described in the issue bring issues around duplicated H1 being added.

A note that Cart page also shares the same problem.

Can we keep the title removal  (disregarding SEO), and add an `aria-label` to the checkout `form` element to ensure screen readers properly announce the page?

cc/ @DaniGuardiola @nikkivias

**Comment by Manussakis** [comment]
@wavvves Thanks for your feedback.

From the accessibility standpoint, having a visible H1 on the page is recommended. However, if Woo decides to go for the UX improvements that the heading removal brings, this is what we need to implement to improve the accessibility of the page:

- Add the `aria-label` attribute to the form element, as you also suggested. This ensures the form is treated as a landmark region, making it visible on the VoiceOver rotor and allowing users to jump directly to its content.
- Remove the `aria-hidden="true"` attribute from all the checkout form subtitles. Otherwise, the checkout page won't have any headings at all. This will also ensure the headings are visible on the VoiceOver rotor.

I have a [PR](https://github.com/woocommerce/woocommerce/pull/50925) in progress and will send it for review soon.

Thanks!
